service/ports/rpc/handler_blobs_get_test.go

```diff
@@ -0,0 +1,154 @@
+	rw := NewMockResponseWriter()
+	req := createBlobsGetRequest(t, id)
+
+	err := h.Handle(ctx, rw, req)
+	require.NoError(t, err)
+
+	require.Len(t, rw.WrittenMessages, 1)
+	require.Equal(t, []byte("test"), rw.WrittenMessages[0])
+
+	queryHandler.RequireThereAreNoOpenReadClosers(t)
+}
+
+func TestLargeBlobIsWrittenToResponseWriter(t *testing.T) {
+	queryHandler := newGetBlobQueryHandlerMock()
+	h := rpc.NewHandlerBlobsGet(queryHandler)
+
+	payloadInFirstMessage := []byte(strings.Repeat("a", rpc.MaxBlobChunkSizeInBytes))
+	payloadInSecondMessage := []byte(strings.Repeat("b", rpc.MaxBlobChunkSizeInBytes))
+	payloadInThirdMessage := []byte("something that doesn't fill up the message completely")
+
+	var payload []byte
+	payload = append(payload, payloadInFirstMessage...)
+	payload = append(payload, payloadInSecondMessage...)
+	payload = append(payload, payloadInThirdMessage...)
+
+	id := fixtures.SomeRefBlob()
+	queryHandler.MockBlob(id, payload)
+
+	ctx := fixtures.TestContext(t)
+	rw := NewMockResponseWriter()
+	req := createBlobsGetRequest(t, id)
+
+	err := h.Handle(ctx, rw, req)
+	require.NoError(t, err)
+
+	require.Len(t, rw.WrittenMessages, 3)
+	require.Equal(t, payloadInFirstMessage, rw.WrittenMessages[0])
+	require.Equal(t, payloadInSecondMessage, rw.WrittenMessages[1])
+	require.Equal(t, payloadInThirdMessage, rw.WrittenMessages[2])
+
+	queryHandler.RequireThereAreNoOpenReadClosers(t)
+}
+
+type getBlobQueryHandlerMock struct {
+	blobs           map[string][]byte
+	openReadClosers map[*readCloserTrackingCloses]struct{}
+}
+
+func newGetBlobQueryHandlerMock() *getBlobQueryHandlerMock {
+	return &getBlobQueryHandlerMock{
+		blobs:           make(map[string][]byte),
+		openReadClosers: make(map[*readCloserTrackingCloses]struct{}),
+	}
+}
+
+func (h getBlobQueryHandlerMock) Handle(query queries.GetBlob) (io.ReadCloser, error) {
+	data, ok := h.blobs[query.Id.String()]
+	if !ok {
+		return nil, errors.New("blob not found")
+	}
+	return newReadCloserTrackingCloses(bytes.NewBuffer(data), h.openReadClosers), nil
+}
+
+func (h getBlobQueryHandlerMock) MockBlob(id refs.Blob, data []byte) {
+	cpy := make([]byte, len(data))
+	copy(cpy, data)
+	h.blobs[id.String()] = cpy
+}
+
+func (h getBlobQueryHandlerMock) RequireThereAreNoOpenReadClosers(t *testing.T) {
+	require.Empty(t, h.openReadClosers)
+}
+
+func createBlobsGetRequest(t *testing.T, id refs.Blob) *transportrpc.Request {
+	args, err := messages.NewBlobsGetArguments(
+		id,
+	)
+	require.NoError(t, err)
+
+	argsBytes, err := args.MarshalJSON()
+	require.NoError(t, err)
+
+	req, err := transportrpc.NewRequest(
+		fixtures.SomeProcedureName(),
+		fixtures.SomeProcedureType(),
+		argsBytes,
+	)
+	require.NoError(t, err)
+
+	return req
+}
+
+type readCloserTrackingCloses struct {
+	reader io.Reader
+	m      map[*readCloserTrackingCloses]struct{}
+}
+
+func newReadCloserTrackingCloses(reader io.Reader, m map[*readCloserTrackingCloses]struct{}) *readCloserTrackingCloses {
```

Comment:
It was hard for me to understand that this map created by the handler is passed here to be changed to effectively notify the handler about closing. I propose to leave the responsibility of modifying the map to just one component - the handler - to make it more obvious. `newReadCloserTrackingCloses` responsibility would be just to call a callback function.

```
func (h getBlobQueryHandlerMock) Handle(query queries.GetBlob) (io.ReadCloser, error) {
	data, ok := h.blobs[query.Id.String()]
	if !ok {
		return nil, errors.New("blob not found")
	}
	
	rc := newReadCloserTrackingCloses(bytes.NewBuffer(data), h.handleReaderCloserClose)
	h.openReadClosers[rc] = struct{}{}

	return rc, nil
}

func (h *getBlobQueryHandlerMock) handleReaderCloserClose(rc *readCloserTrackingCloses) {
	delete(h.openReadClosers, rc)
}

func (r *readCloserTrackingCloses) Close() error {
	r.onClose(r)
	return nil
}
```

Reply:
[9fa4664](https://github.com/planetary-social/go-ssb/pull/10/commits/9fa46648c75c809e9a679aa49ef9ea3db8a99d22)